scrapy/core/http2/protocol.py

```diff
@@ -66,22 +47,45 @@ def __init__(self):
         self.is_connection_made = False
         self._pending_request_stream_pool = []
 
+    def _stream_close_cb(self, stream_id: int):
+        """Called when stream is closed completely
+        """
+        try:
+            del self.streams[stream_id]
+        except KeyError:
+            pass
+
     def _new_stream(self, request: Request):
         """Instantiates a new Stream object
         """
-        stream = Stream(self.next_stream_id, request, self)
+        stream_id = self.next_stream_id
         self.next_stream_id += 2
 
+        stream = Stream(
+            stream_id=stream_id,
+            request=request,
+            connection=self.conn,
+            write_to_transport=self._write_to_transport,
+            cb_close=lambda: self._stream_close_cb(stream_id)
```

Comment:
This may be problematic, when using lambdas `stream_id` may not be what you expect by the time the lambda is executed. See https://docs.python.org/3/faq/programming.html#why-do-lambdas-defined-in-a-loop-with-different-values-all-return-the-same-result

Since the caller of `cb_close` will know the stream ID, you may set `cb_close=self._stream_close_cb` instead and have `Stream` call `self.cb_close(self.stream_id)`.